### DISCUSSION TITLE
How can I make a script to search information in a spreadsheet? [Python]

### DISCUSSION THREAD

**Discussion by nakiahmarie** [community/community#130426] [Programming Help]
### Body

Hi! I am semi comfortable with using python. Currently I am trying to make a script to search for information in a spreadsheet. Column A and B need to be combined into a query, and then searched with the rest of the column headers. Think of google dorking!

I already have ten APIs and ten CSE ids but I am just unsure of what is the best way to approach this. I have tried to use ChatGPT and Gemini to help in both using excel and google scripts but I keep having issues with the code printing out the results correctly. Any help or direction will be appreciated. Thanks!

### Guidelines

- [X] I have read and understood this category's [guidelines](https://github.com/orgs/community/discussions/51384) before making this post.

**Comment by Rishi-Sudhakar** [comment]
Need more info on "Issues with code printing out results"

**Reply by MuhamedHabib** [reply]
** use pandas or pandas ai by reading the spreadsheet pd.read_csv(r"filepath")
** then plan your  search algorithm for example if u want to search for named entities (entreprise name, locations, brands, ....) you could use cosine similarity or fuzzywuzzy 
** turn your notebook into excutable by using pyinstaller
** if you want collaboration try to access  my github account you'll find dozens of notebooks or python application that they include sheets management.

**Comment by HemalathaT2** [comment]
import pandas as pd
import requests
from tabulate import tabulate

# Load the spreadsheet data into a Pandas dataframe
df = pd.read_excel('spreadsheet.xlsx')

# Combine columns A and B into a query string
query = df['A'] + ' ' + df['B']

# Search through the rest of the columns using the query string
results = df[df.columns[2:]].apply(lambda x: x.str.contains(query), axis=0)

# Define the API endpoints and CSE IDs
api_endpoints = ['https://api1.com/search', 'https://api2.com/search', ...]
cse_ids = ['CSE_ID_1', 'CSE_ID_2', ...]

# Loop through the APIs and CSE IDs
for api_endpoint, cse_id in zip(api_endpoints, cse_ids):
    # Send the HTTP request to the API
    response = requests.get(api_endpoint, params={'q': query, 'cx': cse_id})
    # Process the response data
    response_data = response.json()
    # Print the results using tabulate
    print(tabulate(response_data, headers='keys', tablefmt='psql')) pls try this

**Comment by HemalathaT2** [comment]
Step 1: Read the spreadsheet data (use openpyxl or gspread)
Step 2: Combine columns A and B into a query
Step 3: Use the query to search through the rest of the columns
Step 4: Use the APIs and CSE IDs to search
Step 5: Print the results

Have a good day!
